### ISSUE TITLE
[BUG] 2.11.1 Search node disk runs full after some time

### ISSUE THREAD

**Issue by ict-one-nl** [opensearch-project/OpenSearch#11676]
### Describe the bug

We have dedicated search nodes within the cluster with 100GB of SAN-storage as a cache layer. On normal nodes opensearch would have gone red because the watermarks would have been exceeded. No such notice here, it keeps on filling the disks until it is full after which you get errors like:

![image](https://github.com/opensearch-project/OpenSearch/assets/2643715/7c6e4887-2d39-439f-9b51-17a505bc4d96)

![image](https://github.com/opensearch-project/OpenSearch/assets/2643715/a767ddfe-0836-4cb1-800f-be25bad06b62)

We have not set node.search.cache.size because these nods don't do anything other than searching.

### Related component

Storage:Snapshots

### To Reproduce

Leave the cluster running for a while, do several searches to fill the cache -> error.

### Expected behavior

On a dedicated search node use like 90% of the available storage but nothing more. Prevent stability issues.

### Additional Details

**Plugins**
Default docker 2.11.1 images with S3 plugin installed.

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
Kubernetes with docker images 2.11.1

**Additional context**

**Comment by andrross** [comment]
@sandervandegeijn Have you configured a value for [`cluster.filecache.remote_data_ratio`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java#L55-L68)?

Can you share the output of `GET _nodes/stats/file_cache` when a node gets into this state?

**Comment by sandervandegeijn** [comment]
I have implemented the node.search.cache.size setting, setting it to 95GB of the 100GB available on the PVC. I'll see if that holds.

The [cluster.filecache.remote_data_ratio](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java#L55-L68) doesn't do what I need, I don't want to be limited by the cache size. Cache is a bonus and should not prohibit the functioning of the cluster.

**Comment by andrross** [comment]
> The [cluster.filecache.remote_data_ratio](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java#L55-L68) doesn't do what I need if I understand the docs correctly, I don't want to be limited by the cache size.

Unfortunately there are still limits here. If the actively-referenced data at any one time exceeds what is available on the machine then you will run into problems like what you've described here. The result of the `_node/stats/file_cache` API will help determine if that is what is going on here. The `remote_data_ratio` setting is meant as a safeguard to prevent assigning more remote shards that can reasonably be supported.

**Comment by sandervandegeijn** [comment]
Will take a look at it tomorrow. What would reasonable mean, what would be the limiting factor? I'm currently on a cluster with:

```
6 data nodes with 1TB full flash SAN and 16GB RAM
3 master / coordinating nodes and 8GB RAM
10 search nodes with 100GB full flash SAN and 16GB RAM
2 ingest nodesand 2GB RAM
```
There's around 3.5TB on the hot nodes and some 15TB on the S3 cluster. The search nodes makes use of our Scality S3 solution, I'm pulling more than 20gbit/s (and probably more now, implemented more search nodes, no more recent measurements)

**Comment by andrross** [comment]
> What would reasonable mean, what would be the limiting factor?

The strictly limiting factor will be shown by the `file_cache.active_in_bytes`. That is the size of all the actively-referenced (i.e. unevictable) data. If your search workload leads to referencing more data at any point in time than there is storage available on the node, then you'll start seeing failures. In practice, as the `active_in_bytes` size approaches the total cache size then performance might start to degrade as well as you'll see increased cache misses which means that search requests will have to pull down data on-demand more frequently.

**Comment by sandervandegeijn** [comment]
Okay, I've got stats for some of the nodes:

```
"MVyDvVRmT3OP-wzqw2elag": {
      "timestamp": 1703846452872,
      "name": "opensearch-search-nodes-0",
      "transport_address": "10.244.206.182:9300",
      "host": "10.244.206.182",
      "ip": "10.244.206.182:9300",
      "roles": [
        "search"
      ],
      "attributes": {
        "shard_indexing_pressure_enabled": "true"
      },
      "file_cache": {
        "timestamp": 1703846452872,
        "active_in_bytes": 81091561675,
        "total_in_bytes": 0,
        "used_in_bytes": 81703930059,
        "evictions_in_bytes": 265176318281,
        "active_percent": 99,
        "used_percent": 0,
        "hit_count": 313719,
        "miss_count": 63017
      }
    },
    "w1voQkb2SUKuBjDvK_Mfqw": {
      "timestamp": 1703846452883,
      "name": "opensearch-search-nodes-8",
      "transport_address": "10.244.207.169:9300",
      "host": "10.244.207.169",
      "ip": "10.244.207.169:9300",
      "roles": [
        "search"
      ],
      "attributes": {
        "shard_indexing_pressure_enabled": "true"
      },
      "file_cache": {
        "timestamp": 1703846452883,
        "active_in_bytes": 72728861570,
        "total_in_bytes": 0,
        "used_in_bytes": 73525779330,
        "evictions_in_bytes": 257238910529,
        "active_percent": 99,
        "used_percent": 0,
        "hit_count": 349777,
        "miss_count": 59203
      }
    },
    "J3g3YbS7Q_qbvLniDz19RQ": {
      "timestamp": 1703846452883,
      "name": "opensearch-search-nodes-9",
      "transport_address": "10.244.204.127:9300",
      "host": "10.244.204.127",
      "ip": "10.244.204.127:9300",
      "roles": [
        "search"
      ],
      "attributes": {
        "shard_indexing_pressure_enabled": "true"
      },
      "file_cache": {
        "timestamp": 1703846452883,
        "active_in_bytes": 75778946930,
        "total_in_bytes": 0,
        "used_in_bytes": 76533921650,
        "evictions_in_bytes": 258573952202,
        "active_percent": 99,
        "used_percent": 0,
        "hit_count": 362107,
        "miss_count": 60526
      }
    }
  }
```

From what you've written, is the data always transferred to the local storage of the nodes? I was under the impression it would get the data from S3 and it would write the most used parts to the local storage for performance gains. That there will be more cache misses when the local storage / S3 ratio changes is obvious (and a performance penalty I would happily pay to be able to archive the data).

Since I've implemented node.search.cache.size at effectively 95% of the local storage I've not seen errors anymore (so far).

**Comment by sandervandegeijn** [comment]
Okay, had another run in with full disks on the search nodes. From a sysadmin's perspective, the cache should be a bonus for performance optimisation and should not fill up the entire disk of the search node regardless of the total storage in use on the S3 side.

Can I make this happen? :)

**Comment by andrross** [comment]
> I was under the impression it would get the data from S3 and it would write the most used parts to the local storage for performance gains...From a sysadmin's perspective, the cache should be a bonus for performance optimisation and should not fill up the entire disk of the search node regardless of the total storage in use on the S3 side.

@sandervandegeijn Fair points! However, the current architecture requires that the blocks pulled from the remote object store be written to disk, and then open as files on disk to serve searches. This means we don't have to have all actively-referenced data loaded into memory because we can rely on the operating system to page this data between the disk and page cache as appropriate.

I believe the scenario you are encountering is that the sum of the actively-referenced data is exceeding the disk space on the node. At a minimum, I believe a better experience would be to fail new searches that come it that require downloading new data, as opposed to letting the system run out of disk space and fail unpredictably. What do you think?

Separate, it does look like there's another bug here where your file cache stats are showing `"total_in_bytes": 0`. I've not seen that before but "total_in_bytes" is supposed to be the size of the file cache and it is clearly not zero here.

**Comment by sandervandegeijn** [comment]
Okay, uhm... This kind of things are better to discuss in person and with a whiteboard, but I'll try;

Failing gracefully and handling the situation is always better than just letting it blow up. But in this case it's not even doing searches in parallel, this is one search that's going through ±20TB of data where 3.5TB is on full flash storage (within the cluster) and the rest on S3. The search nodes (10) have full flash storage (100GB each), so 1TB of active storage in addition to the hot nodes. It explains why I'm seeing 20-25gbit/s from the S3 storage: basically it's pulling *all* data in as fast as possible. It also explains why doing concurrent searches over the whole dataset is almost impossible. 

Without having any knowledge of the caveats and I mean this as constructive feedback

This 1:20 ratio is about as much as I can stomach in terms of storage costs, ideally I would like something 1:100 where I can live with slower performance. I really don't care that it would take 5-10 minutes for 1 search, it's archive, cold cold storage. Lower ratios negate the advantage of the lower storage costs on the S3 side and it's not going to be practical. I'm also talking to another university that wants to store & search 5 petabyte (!) of logs. Searching 5PB this way seems almost impossible, even when you would scale to multiple 100gbit interlinks. You need too much primary storage to be cost effective when the ratio's are too low. 

That's a shame, it would be the ideal cost effective solution to be able to do this. Thinking through this, it seems we have two problems:

1. All referenced data needs to be on disk, so large searches also require large amounts of primary storage in the cluster, maybe almost in the 1:<10 ratio.
2. It needs to pull all the data from the object storage

Are there ways to improve the implementation in such a way that not always all storage needs to be copied, but data can just be used directly from the object storage and where the cache in the search nodes prevent excessive reads on the most used data?

If this is not possible, I'm thinking about totally different ways of scaling - maybe doing one primary cluster for hot ingest and then a cluster per time unit (month?) with slower storage while using cross cluster search to search through everything at once. 

Have you experimented with these kinds of cluster for petabytes of data?

**Comment by andrross** [comment]
Thanks for the feedback! To clarify, and to get deeper into the technical details...

Searchable snapshots use the Lucene Directory and IndexInput interfaces to abstract away the remote versus local details. When an IndexInput instance is opened, if the specific part of the file being pointed to is not already on disk, it pulls down that part of the file in an 8MiB chunk. When the "cursor" of the IndexInput moves past the current 8MiB part it will then download the next 8MiB part, keeping previous parts cached on disk. So in the absolute simplest case, if you had a single search on a single shard that had a single open IndexInput, then only 8MiB of data would be actively referenced at any one time. Reality is generally more complicated than that, but I would not expect a single search to have to actively reference all data of the index. I think we need to dig deeper into specifically what is happening in this case, because it does not match my expectation for how the system should perform.

**Comment by sandervandegeijn** [comment]
Thanks for the explanation. Would it be necessary to write those chunks to disk or would it be possible to do this in memory and either discard the data immediately or write it to disk when it's a block that used frequently as a cache? The current problem is that - from what I understand - when you do a search over a large dataset a large portion of that dataset needs to be on primary storage. 

For us, the point of using searchable snapshots like this is to save on storage costs (full flash vs S3 is about 20x). This would enable us to keep data available for 180 days (security forensics use case).

I'm happy to provide details or schedule a call to show you around the cluster.

**Comment by sandervandegeijn** [comment]
Okay, unfortunately I'm having to party scrap the current setup. The search node problems cause too many problems, cluster went down again because of it. I'll try to leave the data on the S3 side and maybe even take snapshots so I can restore everything in a second cluster to be able to help debug. But this is getting too cumbersome for production uses I'm afraid.

**Comment by andrross** [comment]
> The current problem is that - from what I understand - when you do a search over a large dataset a large portion of that dataset needs to be on primary storage.

@sandervandegeijn This should not be the case. With a very large data set and a relatively small primary storage cache, the expected behavior is that you might see a lot of thrashing - i.e. re-downloading of data that was evicted from primary storage - but that should manifest as slow queries and not cluster instability or disk exhaustion. Obviously something is not working correctly here though.

**Comment by sandervandegeijn** [comment]
Okay is what I would have expected as well. With a smaller cache, cache misses will be more frequent and more traffic will be generated. That's to be expected.

**Comment by sandervandegeijn** [comment]
I had to delete the data as well (for now). Do you have any means to reproduce this at scale?

**Comment by peternied** [comment]
[Triage - attendees [1](https://github.com/peternied) [2](https://github.com/lukas-vlcek/) [3](https://github.com/andrross) [4](https://github.com/CEHENKLE) [5](https://github.com/dbwiddis) [6](https://github.com/reta/) [7](https://github.com/anasalkouz) [8](https://github.com/aggarwalShivani)]
@sandervandegeijn Thanks for filing this issue

**Comment by sandervandegeijn** [comment]
Is it possible to solve this in the 2.12 release?

**Comment by sandervandegeijn** [comment]
Guys, is there any chance this will be solved in 2.13? I can't use the future right now because it makes the clusters unstable.

**Comment by peternied** [comment]
@sandervandegeijn what do you think about initiating a draft pull request to tackle this challenge, or perhaps motivating someone directly impacted by it to contribute a pull request?  I'd be happy to help review the changes.

**Comment by sandervandegeijn** [comment]
Afraid I'm not a very capable java dev, so it's quite a task for me. Second excuse: I can't build opensearch on my macbook as far as I can tell (which is a requirement for the coming weeks, not taking the Lenovo P15 hernia inducing brick to the camping ;) )

**Comment by peternied** [comment]
@sandervandegeijn Hmm I thought ARM was supported for core, but I'm a windows person that ssh's into linux machines 🤷.  If you have time give it a try and feel free to create issues if it doesn't work.  Maybe you know someone at your university that could take on the challenge while you are away.  Enjoy camping in any case!

**Comment by sandervandegeijn** [comment]
Promise I'll give it a try, but I hope it won't come to my java skills to get this fixed for 2.14 ghehe.

**Comment by Pallavi-AWS** [comment]
@anasalkouz can we please look into prioritizing handling disk issues with small disk cache compared to the remote data being addressed?

**Comment by andrross** [comment]
First step here should be to replicate the problem by configuring a small disk cache relative to the remote data being addressed. We should see the cache being over-subscribed (and in the worst case it can result in completely filling the storage device containing the file cache). Once we have a reproduction in a controlled environment, the avenues to explore/consider that I can think of are:

- Can we fail search requests in lieu of over-subscribing the local storage?  This isn't a great user experience but it is much better than node/cluster instability.
- Are we possibly leaking file references somewhere? We have a reference counting mechanism to determine if cached files are unused, but it is always possible there is a bug in that logic.
- Can we be more aggressive about cleaning up unreferenced cache files? Because Lucene doesn't directly close IndexInput instances it creates via the `clone()` method, we rely on a [GC cleaner](https://github.com/opensearch-project/OpenSearch/blob/c0852f84f0647f78a45e71396d6370a39ba4922f/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java#L90) to close those instances. That means that potentially unreferenced files may still be counted as referenced until the garbage collector decides to collect the instance. GC is generally pretty lazy about cleaning up things when it still has available heap, which could be an aggravating problem here.
- The node stats output above shows `"total_in_bytes": 0`. That field should be the total configured size of the cache, so this certainly looks like a bug. While potentially innocuous, we should get a reproduction and fix this and determine if it is in anyway related to the larger problem observed here.

Any other ideas are welcome as well.

**Comment by sandervandegeijn** [comment]
Thanks Andrew, good to have some avenues to explore. I'm busy on building a formal test environment of our setup this week and next week (rebuilding the CI/CD as well, so it's going to take a little more effort than usual). 

When that's done I can assist with testing without disrupting our production workflows. Don't hesitate to ask :)

**Comment by anasalkouz** [comment]
Thanks Andrew, looks we have some steps/actions we can do to move this forward. Thanks Sundar, I think having testing environment will definitely help to easily reproduce and effectively test any fixes. Hopefully, we can have someone to look into the issue in the next couple of weeks.

**Comment by sandervandegeijn** [comment]
Hi guys, 2.14 is right around the corner. Any chance this will be resolved in 2.14? Thanks!

**Comment by rlevytskyi** [comment]
> (and in the worst case it can result in completely filling the storage device containing the file cache)

I have this issue with 2.12.

Here is allocation report:
`% curl logs:9200/_cat/allocation\?v
shards disk.indices disk.used disk.avail disk.total disk.percent host                        ip           node
   283      287.3gb     4.7gb     59.1gb     63.9gb            7 v87-search.co.com 10.139.20.28 v87.co.com
  2319        2.2tb     2.3tb      1.1tb      3.4tb           66 v81-data.co.com   10.139.20.11 v81-data.co.com
...`

At the same time, OS at VM report:
`v87:[/opt/os-search]# df -h /mnt/search/
Filesystem      Size  Used Avail Use% Mounted on
/dev/xvdc1       64G   60G  4.3G  94% /mnt/search`

Here is cache size limit:
`v87:[/opt/os-search]# grep cache config/opensearch.yml 
node.search.cache.size: 20gb`

**Comment by sandervandegeijn** [comment]
@andrross that looks like the underlying problem! Nice, is it easy to fix?

**Comment by sandervandegeijn** [comment]
Unfortunately we have missed 2.14, (how) can we get this prioritized for 2.15?

**Comment by andrross** [comment]
We do have a link to this issue in the known limitations section here: https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/#known-limitations

**Comment by sandervandegeijn** [comment]
Is the documentation correct? I mean that it's probably always be a problem, but the chance of it occurring increases when the ratio is larger?

I can rephrase that but only if it's valid :)

**Comment by finnegancarroll** [comment]
Working on this issue!

**Comment by evheniyt** [comment]
We have the same issue in our cluster with 35Tb of data on S3 and 3 running search nodes with 650Gb of disk space each. We set `node.search.cache.size: 200Gb` for search nodes.
We could help with reproducing this on a test environment or testing the fix.

**Comment by damslo** [comment]
> We have the same issue in our cluster with 35Tb of data on S3 and 3 running search nodes with 650Gb of disk space each. We set `node.search.cache.size: 200Gb` for search nodes. We could help with reproducing this on a test environment or testing the fix.

I have almost the same issue, but we even set cache to 2tb

**Comment by finnegancarroll** [comment]
So far I’m not seeing cases where a search node is completely leaking file references. It is however, very easy to over subscribe the local file cache while addressing a sufficiently large remote snapshot. The two ways I notice this happening are as follows:  

1.) As discussed above Lucene does not promise to close IndexInputs(file references) after completing a search. Opensearch relies on GC to close these after the fact, which can allow references to overstay their welcome if GC is not aggressive enough.

2.) To reduce file handle usage Lucene collapses several per segment files into a single [Compound File](https://lucene.apache.org/core/2_9_4/fileformats.html#Compound%20Files). These files appear to maintain high reference counts regardless of GC and as such are never evicted even as the cache overflows.

If possible the below information would help narrow down the issue:

- Forcing (more like “suggesting”) GC on an oversubscribed search node consuming excessive disk space and observing to see if disk space is freed as a result.
`jcmd <pid> GC.run`

- I would be interested to see how many segments are present in indices serviced by the search node.
`/<index>/_stats?pretty`

- To verify that disk space used is actively accounted for by the file cache.
`/_nodes/stats/file_cache?pretty`

PS - In the case of `"total_in_bytes": 0`, Opensearch will probe the 'path.data' directory with [getUsableSpace()](https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/nio/file/FileStore.html#getUsableSpace--) as an estimate. This is liable to return 0 if there is any external disk I/O going on concurrently. I would be interested to know if this bug is consistently reproducible as that would eliminate the possibility of some background disk operation messing with this method.

**Comment by sandervandegeijn** [comment]
I don't have a formal testing env I can use for the tests, I'm hesitant to do it in production since the last times it blew up. I'm also somewhat pressed for time because our daughter was born last week. I can help in a couple of weeks.

Is there someone else that could help out?

**Comment by sandervandegeijn** [comment]
Hi is there any news? I'm starting at work again, I can see if I can help out. :)

**Comment by rlevytskyi** [comment]
No news.
Issue still persists at 2.12

**Comment by sandervandegeijn** [comment]
Without changes I'm hesitant to try again (and possibly blow up another cluster ghehe). We have missed the 2.15 release window, I'm hoping for 2.16. This issue is important to opensearch being a valid candidate for multiple educational institutions in the Netherlands. We need the storage tiering this feature provides to keep the costs acceptable.

**Comment by finnegancarroll** [comment]
This issue is still being investigated and I'll be spending time this week working towards a fix.
Can you share more details of your use case? In particular how much query throughput the affected nodes experience and roughly how long it takes for a node to fill its disk?

**Comment by sandervandegeijn** [comment]
Great thanks!

In our case we had 10 search nodes with 100GB primary storage each with about 20TB of snapshot data. It took 10-20 searches searches to break the cluster, searches being hours or days apart. During the search I was peaking at about 25gbit/s of throughput.

**Comment by andrross** [comment]
> about 20TB of snapshot data

@sandervandegeijn Can you share some statistics about this data? I think it might be helpful to know: total number of indexes, total number of shards, average number of segments per shard, and number of replicas (if any) configured for the searchable snapshot indexes.

**Comment by sandervandegeijn** [comment]
Okay, I'm typing along while setting up the testing environment.

First conservative try, 5 search nodes, 20GB storage each. Couldn't restore the 30TB-ish of data, exceeded max number of shards per node. Also, the primary storage of the search nodes is 100% used without firing one query. 

Retrying with 10 search nodes, 100GB of storage each and 10k shards per node max. No replicas configured.

**Comment by finnegancarroll** [comment]
The incorrect reporting of disk space could be related to this recently merged issue: #14004.
When restarting a search node with an existing file cache already occupying disk the calculated `availableCapacity` does not account for files already in the cache. `availableCapacity` is then reported as the difference between the total remaining space and previous cache size.

**Comment by sandervandegeijn** [comment]
If you have a docker image for me I can verify? :)

**Comment by finnegancarroll** [comment]
To clarify, I don't think this fix will impact your actual disk usage. While the capacity of the file cache is incorrectly reported in some cases according to #14004, the file cache capacity is only used to determine if the cache is overflowing and can begin evicting stale entries. So an artificially low capacity would only make evictions start sooner.

**Comment by finnegancarroll** [comment]
I believe the fastest way to reduce disk usage will be to make the [size of fetched blocks](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java#L384) configurable to the user which I am working on [here](https://github.com/opensearch-project/OpenSearch/pull/14753).

The current default block size of 8Mib is quite large considering our cache is largely populated by metadata when restoring a remote searchable snapshot and it seems as if we are agnostic as to the size of the data we actually need to perform any given operation. That is, 100bytes of required metadata in our remote repository will likely be downloaded as an 8Mib block.

By modifying my block size to be smaller i've seen the disk usage of my file cache drop drastically.

**Comment by sandervandegeijn** [comment]
Cool! Is that something that is adjustable after snapshots have been created already? Or can it be adjusted for a current repo for new snapshots.

But from what I've read this makes the odds of this bug occurring smaller, but it doesn't fix the underlying problem, is that correct? Can I provide more info to help find a structural fix for this one?

**Comment by sandervandegeijn** [comment]
Any news? Can I help you guys somehow? I'm afraid we're going to miss the 2.16 deadline.

We are moving along with our plans for a new SOC design, opensearch can fill an important role here, but we kind of need this functionality. Another option would be to tier the storage within the cluster (hot/warm architecture), but we like the data being immutable as a snapshot and it brings less overhead.

**Comment by linuxpi** [comment]
[Storage Triage - attendees [1](https://github.com/ashking94/) [2](https://github.com/shourya035/) [3](https://github.com/gbbafna/) [4](https://github.com/andrross/) [5](https://github.com/ltaragi/) [6](https://github.com/astute-decipher/) [7](https://github.com/skumawat2025/) [8](https://github.com/linuxpi/)]

@finnegancarroll Do you think we can share some information here or can @sandervandegeijn help provide some more details which might help. Lets see if we can target for 2.17 maybe?

**Comment by finnegancarroll** [comment]
> Cool! Is that something that is adjustable after snapshots have been created already? Or can it be adjusted for a current repo for new snapshots.
> 

As planned the effect of this block size setting would be limited to the local cache so I think it would be appropriate as a node level setting. The remote snapshot would not need to change.
 
> Will this incur more calls to the s3 storage when you make it smaller? Then there might an optimum to be found.
>

This is definitely a con of lowering the block size and i've noticed performance hits when pushing the block size down too far.

> But from what I've read this makes the odds of this bug occurring smaller, but it doesn't fix the underlying problem, is that correct? Can I provide more info to help find a structural fix for this one?

That's correct. This change will not fix the underlying problems with this cache and is only an attempt to soften the impact on disk usage. I have not been able to produce a case where subsequent queries inflate the size of the cache indefinitely. What I have noticed is the configured `capacity` of this particular cache has little bearing on it's disk usage. The logic which removes cached files from disk is tightly coupled with garbage collection causing heavy spikes in disk usage up until GC runs and the cache can be "cleaned". With this change I would only hope to shrink these spikes, a true fix will likely require a change in design.

**Comment by finnegancarroll** [comment]
Hi @linuxpi,

Here are some more details regarding the behavior of this cache as I currently understand it.

Opensearch provides Lucene [OnDemandBlockSnapshotIndexInputs](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInput.java) which negotiate downloading blocks from the remote store while implementing all the functionality of a regular IndexInput. This is accomplished by downloading whichever 8MB block of the file is referenced by the IndexInput whenever the data is actually needed by Lucene.

Each block downloaded by an OnDemandBlockSnapshotIndexInput is [stored in a ref counted file cache](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java#L71). The originating IndexInput does not contribute to the reference count but all [clones of that block increment the reference count](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/filecache/FileCachedIndexInput.java#L129). The reference count for any given block is decremented when it is closed. This is problematic because Lucene makes no promise to close IndexInputs it has cloned. The solution to this is where our design creates a dependency on GC. Each IndexInput is [registered](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java#L90) with a [java Cleaner](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) which maintains a list of phantom references to each IndexInput. When GC runs any object in the cleaner's list that is now only phantom reachable is moved into a queue. A single thread is constantly iterating through this queue and calling the cleanup task for each object, in our case, [closing the IndexInput](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/filecache/FileCachedIndexInput.java#L141).

This design not only ties our disk usage to the performance and behavior of GC, but also the single thread registered with the cleaner that must actually do the work of closing our IndexInputs, then decrementing their ref count. It's also worth noting that even when GC runs you will often be left with a full cache despite reaching a reference count of zero on many of your entries. For example:

Directly after forcing GC
```
"file_cache" : {
    "timestamp" : 1722210969134,
    "active_in_bytes" : 249869432,
    "total_in_bytes" : 10485760,
    "used_in_bytes" : 434418808,
    "evictions_in_bytes" : 51220894703,
    "active_percent" : 58,
    "used_percent" : 4143,
    "hit_count" : 57169,
    "miss_count" : 6221
}
```

While GC might close all phantom reachable clones of an IndexInput, the actual cache will not trigger an [eviction](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java#L374) and close the originating IndexInput until it is accessed. So you often have IndexInputs with ref count zero remaining on disk while Opensearch is idle.

The effect of this seems to be disk usage that trends upwards during a search, with substantial drops that correlate with GC followed closely by a cache eviction.

Here's a small snippet of my cache size in bytes logged every second while running the OSB big5 workload from a remote snapshot: [directory_size_bytes.log](https://github.com/user-attachments/files/16407065/directory_size_bytes.log)

In addition to the above Lucene does its own caching and holds onto some select IndexInputs containing metadata for each segment in an index. These IndexInputs maintain a high reference count for the lifetime of the cache and can never be evicted. #14990 has some information relating to how that behavior effects the baseline size of our cache.

**Comment by finnegancarroll** [comment]
In terms of structural changes I've been wondering if we could move to a true LRU cache with strictly enforced capacity. It seems like we could extend our cache implementation to evict blocks from disk without waiting for their ref count to reach zero as long as we hold onto the metadata necessary to fetch the block once it's needed again.

**Comment by andrross** [comment]
@finnegancarroll What about changing the behavior to fail a search instead of allowing the cache size to be oversubscribed? That doesn't necessarily solve the underlying problem but it is a better experience than allowing the node to crash.

**Comment by sandervandegeijn** [comment]
That would be better, better fail to gracefully (kind of). The consequences are a lot easier to deal with than a red cluster that doesn't ingest data anymore.

> In terms of structural changes I've been wondering if we could move to a true LRU cache with strictly enforced capacity. It seems like we could extend our cache implementation to evict blocks from disk without waiting for their ref count to reach zero as long as we hold onto the metadata necessary to fetch the block once it's needed again.

That does sound like a better solution, at least stability would be guaranteed. Maybe there are more intelligent schemes than LRU even.

That being said, the ideal solution would not require a cache at all. To be really nit-picky: if it's required it's not really a cache. The cache should be there to optimize performance, having less of it should only mean less/degraded performance. If this would be possible, stability would be almost guaranteed.

**Comment by sandervandegeijn** [comment]
2.16 is out now, what can we do to resolve this for the 2.17.0 release?

**Comment by AmiStrn** [comment]
We may benefit from a cache clearing API to mitigate this issue when we get rejected queries or when the cache is suspected of overflowing. Currently, the cache clearing API does not clear the cache as it does with other caches (like the request/fielddata caches)

**Comment by sandervandegeijn** [comment]
Sorry I'm triggering a bit on the 6 months plus label, this bug is already 8 (!) months old and breaks production functionality. If little priority is given to this, that would be a shame but I can't oversee the value of other issues. That might be valid.

But in that case searchable snapshots should be removed as a functionality until it is fixed because it's unreliable and it's not a matter if it will break a cluster but when. This should not be in a stable opensearch release.

**Comment by sandervandegeijn** [comment]
I'm afraid I can't validate the fix, but I am pushing the storage ratio to the extremes (300GB primary on 80TB of S3):

![image](https://github.com/user-attachments/assets/5f2d5a80-365b-4d82-8738-91f0ad630846)

Result: red cluster

**Comment by sandervandegeijn** [comment]
I'm afraid I can't validate the fix, but I am pushing the storage ratio to the extremes (300GB primary on 80TB of S3):

![image](https://github.com/user-attachments/assets/117f7446-66af-4930-889e-9adfa62368c0)


Result: red cluster

**Comment by drewmiranda-gl** [comment]
For what its worth I can't find `cluster.filecache.size` referenced anywhere on github or in the OpenSearch docs. Do you have any reference information about it?

**Comment by ddijk** [comment]
Ah, you're right, `cluster.filecache.size` is only available in Elastic Search.

**Comment by ddijk** [comment]
Looks like it would be a good idea to implement `cluster.filecache.size` for OpenSearch. Would love to if I can find the time. Agreed anyone or other suggestions?

**Comment by sandervandegeijn** [comment]
That would be great, but would that limit the effective cache size? The unpredictable evictions seem to be the problem if I understand Andrew correctly.

**Comment by HenryTheSir** [comment]
Hi @sandervandegeijn ,

what is the replica count of your remote store indices? (Sorry if I missed that part somewhere up)

Best regards

**Comment by sandervandegeijn** [comment]
0 :) np!

**Comment by damslo** [comment]
Hello,
Anyone tested this issue on Opensearch 3.0?

**Comment by sandervandegeijn** [comment]
Nope, I'm waiting a few minor releases before upgrading to 3.0 to iron out the initial bugs.

We are running with searchable snapshots now, but with more storage in the search nodes. The metadata was eating too much of the disk space initially. With 1TB on 40TB of S3 it's fine for now.

**Comment by thomas315** [comment]
I can confirm that this is also happening in version 3.10

What I observed so far is that `node.search.cache.size` must be set higher than the storage needed for initializing all searchable snapshots. That's the only limiting factor. Query searches after that are exceeding the value set by this property and if disk size is too small it will fill the whole disk.
Setting `remote_data_ratio` doesn't do anything for me, but maybe I'm setting it wrong. I tried ratio of 1 and 5. But each time the behavior was the same:
S3 data: 25GB
Needed for restoring them: ~12GB
Disk size 2x 12GB
`node.search.cache.size` not set --> 80% of Disk Space (~10GB see stats below)

Query searches filling up the disk slowly until it's full. Since this took dozens of searches I would expect that it evicts enough old data at some point and not fail the search.
Tests were performed only with this small dataset for now, will be interesting how it performs with more data.

So it seems the only solution is to have enough disk storage to satisfy all possible searches.
Would you mind describing what exactly `remote_data_ratio` can do? What does a value of 5 actually do? Unfortunately this wasn't clear enough to me in the docs.

Stats:
```
# Before hitting full disk
      "aggregate_file_cache": {
        "timestamp": 1755235265248,
        "active_in_bytes": 6517829583,
        "total_in_bytes": 10058891264,
        "used_in_bytes": 10047940063,
        "pinned_in_bytes": 0,
        "evictions_in_bytes": 2060574729,
        "active_percent": 65,
        "used_percent": 100,
        "hit_count": 197292,
        "miss_count": 23132,
        "over_all_stats": {
          "active_in_bytes": 6517829583,
          "used_in_bytes": 0,
          "evictions_in_bytes": 2060574729,
          "active_percent": 65,
          "hit_count": 197292
        },

# Just after hitting full disk (12GB) and search fails
      "aggregate_file_cache": {
        "timestamp": 1755235438262,
        "active_in_bytes": 6635270095,
        "total_in_bytes": 10058891264,
        "used_in_bytes": 10028763001,
        "pinned_in_bytes": 0,
        "evictions_in_bytes": 2448850543,
        "active_percent": 66,
        "used_percent": 100,
        "hit_count": 255260,
        "miss_count": 23178,
        "over_all_stats": {
          "active_in_bytes": 6635270095,
          "used_in_bytes": 0,
          "evictions_in_bytes": 2448850543,
          "active_percent": 66,
          "hit_count": 255260
        },

```

**Comment by sandervandegeijn** [comment]
The implementation seems to be worse in 3.x

I have set the file cache size to 450gb, opensearch reports:  

`opensearch-search-nodes [2025-11-01T10:56:45,275][INFO ][o.o.n.Node               ] [opensearch-search-nodes-1] cache size [450gb]`

So it does see the setting. It tries to initialize all the indices, but it stops at a certain point:

```
 {
      "node_id": "58s2V0wUQL-QaaQnpsFkhw",
      "node_name": "opensearch-search-nodes-1",
      "transport_address": "10.244.130.106:9300",
      "node_attributes": {
        "shard_indexing_pressure_enabled": "true"
      },
      "node_decision": "no",
      "weight_ranking": 1,
      "deciders": [
        {
          "decider": "max_retry",
          "decision": "NO",
          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2025-11-01T10:29:31.680Z], failed_attempts[5], failed_nodes[[S47Gp9FMT2Gka9xourrcXA]], delayed=false, details[failed shard on node [S47Gp9FMT2Gka9xourrcXA]: failed recovery, failure RecoveryFailedException[[log-suricata-ike-2025.week-13-snapshot][0]: Recovery failed on {opensearch-search-nodes-0}{S47Gp9FMT2Gka9xourrcXA}{KCzFK-hsRO-0akTZDPODFA}{10.244.130.102}{10.244.130.102:9300}{w}{shard_indexing_pressure_enabled=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: NotSerializableExceptionWrapper[unchecked_i_o_exception: java.io.IOException: Local file cache capacity (83886080000) exceeded (83894228881) - BlobFetchRequest failed: /usr/share/opensearch/data/nodes/0/cache/14Gp28BWSJOTyaSjYhq26A/0/RemoteLocalStore/_xf7_Lucene912_0.doc_block_0]; nested: IOException[Local file cache capacity (83886080000) exceeded (83894228881) - BlobFetchRequest failed: /usr/share/opensearch/data/nodes/0/cache/14Gp28BWSJOTyaSjYhq26A/0/RemoteLocalStore/_xf7_Lucene912_0.doc_block_0]; ], allocation_status[deciders_no]]]"
        }
      ]
    },
    {
      "node_id": "pcHtKV4VSUmyNS0DU4U5NA",
      "node_name": "opensearch-search-nodes-0",
      "transport_address": "10.244.33.19:9300",
      "node_attributes": {
        "shard_indexing_pressure_enabled": "true"
      },
      "node_decision": "no",
      "weight_ranking": 2,
      "deciders": [
        {
          "decider": "max_retry",
          "decision": "NO",
          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2025-11-01T10:29:31.680Z], failed_attempts[5], failed_nodes[[S47Gp9FMT2Gka9xourrcXA]], delayed=false, details[failed shard on node [S47Gp9FMT2Gka9xourrcXA]: failed recovery, failure RecoveryFailedException[[log-suricata-ike-2025.week-13-snapshot][0]: Recovery failed on {opensearch-search-nodes-0}{S47Gp9FMT2Gka9xourrcXA}{KCzFK-hsRO-0akTZDPODFA}{10.244.130.102}{10.244.130.102:9300}{w}{shard_indexing_pressure_enabled=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: NotSerializableExceptionWrapper[unchecked_i_o_exception: java.io.IOException: Local file cache capacity (83886080000) exceeded (83894228881) - BlobFetchRequest failed: /usr/share/opensearch/data/nodes/0/cache/14Gp28BWSJOTyaSjYhq26A/0/RemoteLocalStore/_xf7_Lucene912_0.doc_block_0]; nested: IOException[Local file cache capacity (83886080000) exceeded (83894228881) - BlobFetchRequest failed: /usr/share/opensearch/data/nodes/0/cache/14Gp28BWSJOTyaSjYhq26A/0/RemoteLocalStore/_xf7_Lucene912_0.doc_block_0]; ], allocation_status[deciders_no]]]"
        }
      ]
    }
```

```
GET _nodes/opensearch-search-nodes-0/settings?filter_path=**.node.search.cache.size

{
  "nodes": {
    "pcHtKV4VSUmyNS0DU4U5NA": {
      "settings": {
        "node": {
          "search": {
            "cache": {
              "size": "450gb"
            }
          }
        }
      }
    }
  }
}
```

Where is that 80GB coming from?

<img width="647" height="236" alt="Image" src="https://github.com/user-attachments/assets/dca5561e-6611-4c8a-898c-d5ae8c134829" />

<img width="616" height="229" alt="Image" src="https://github.com/user-attachments/assets/90127a02-cdb9-4041-b989-05f913418e06" />

<img width="892" height="424" alt="Image" src="https://github.com/user-attachments/assets/31af8d0f-e7ba-4291-813c-47b3f9a31b8f" />

Disk usage is off and seems to be influenced by the remote data ratio? (which should only be a check).

I'm seriously considering going back to 2.x despite the data loss that would occur. It wasn't perfect by a long shot, but better than this.

**Comment by mishail** [comment]
@sandervandegeijn while I understand you may have valid reasons to be frustrated, can you edit your comment to remove explicit language? Thanks!

**Comment by sandervandegeijn** [comment]
Sure, sorry. Been at this for almost two years and spent a couple of grand trying to get it solved. Now I'm even getting null pointer exceptions on restoring (submitted another issue for that).

We might need to consider pulling this from stable IMHO.

----
I might(!!) have found the source of the error in my previous post, it lost the ratio setting it seems. But the error is very vague if that's the cause. But since I'm not able to restore any of the snapshots because of the null pointer I'm not able to verify.

**Comment by mburni** [comment]
Hi,

I see that even mounting a searchable snapshot consumes some memory in the cache. When cache is full, mounting additional snapshot is impossible and the cluster becomes "red" because of partially mounted snapshots. Is that the intended behavior?

@sandervandegeijn do I get you right? Your cluster gets "red" when running out of disk space in the search cache and you're not able to ingest new data? So, you break ingestion by searching?

Can someone please tell me what the impact of a full search cache with "dedicated search nodes" will be:

- Failed searches?
- No ingestion of new data?
- Failed OpenSearch-process?

My environment:

- OpenSearch 2.19.3
- 42 nodes, 2 of them are search nodes 
- Planned: 4000 searchable snapshots, 240tb in snapshot repository

B.t.w.: I also see "null pointer exceptions" when trying to mount the searchable snapshots. I work around by sleep about 40 seconds between the api-calls.

**Comment by sandervandegeijn** [comment]
Okay, I'm up and running again, sort of. The restore problem caused by the null pointer exception has been solved, restoring snapshots works again. 

We have effectively disabled the ratio setting by setting it to 1000. In practice this works okay now given that you provide enough spare room for the initial cache en subsequent searches. 100T of S3 with 1T of primary storage with infrequent use of the data on S3 is fine for us, the growth of the cache is limited. One search node has 72% disk usage, the other one 83%. So pushing further than 1:100 will be problematic because the initial cache/metadata for all indices consume so much storage space. 

I might provision two more search nodes to get more headroom and it would give us more I/O towards our storage cluster that hosts the S3.

When it goes wrong it depends, but I have had a red cluster that didn't ingest anymore.

**Comment by sandervandegeijn** [comment]
With the help of Claude I have a way to reproduce: https://github.com/sandervandegeijn/opensearch-issue-11676/tree/main

This looks like the behavior I'm seeing in production but with a lot less data. Just toying around in https://github.com/sandervandegeijn/OpenSearch :)

**Comment by sandervandegeijn** [comment]
Like I indicated in the merge thread: fantastic. Many many thanks! This solves one of our main problems with Opensearch :)

**Comment by mvukadinoff** [comment]
Thanks for the great work on this. Is there any chance this will be backported to Opensearch 2.XX latest patch version? We're using a software (Graylog) which is not officially supporting version 3.X
